File: Python/Problems/Problems/problem3471.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem3471(Problem):
    def test(self) -> bool:
        nums = [3, 9, 2, 1, 7]
        k = 3

        expected = 7

        result = self.largestInteger(nums, k)

        print(result)

        return result == expected

    def largestInteger(self, nums: List[int], k: int) -> int:
        n = len(nums)
        counts = [0] * 51
        for num in nums:
            counts[num] += 1

        if k == 1:
            for i in range(50, -1, -1):
                if counts[i] == 1:
                    return i
        elif k == n:
            for i in range(50, -1, -1):
                if counts[i] != 0:
                    return i
        else:
            for i in range(50, -1, -1):
                if counts[i] == 1 and (i == nums[0] or i == nums[n - 1]):
                    return i

        return -1

File: C++/Problems/Problems/problem_3471.cpp
#include "problem_3471.h"

bool problem_3471::test() {
    std::vector nums { 3, 9, 2, 1, 7 };
    int k = 3;

    int expected = 7;

    auto result = largestInteger(nums, k);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_3471::largestInteger(std::vector<int> &nums, int k) {
    int n = nums.size();
    std::vector counts(51, 0);
    for (auto num : nums)
        counts[num]++;
    if (k == 1)
    {
        for (int i = 50; i >= 0; --i)
            if (counts[i] == 1)
                return i;
    }
    else if (k == n)
    {
        for (int i = 50; i >= 0; --i)
            if (counts[i] != 0)
                return i;
    }
    else
    {
        for (int i = 50; i >= 0; --i)
            if (counts[i] == 1 && (i == nums[0] || i == nums[n - 1]))
                return i;
    }
    return -1;
}
